src/Aspire.Hosting/ApplicationModel/EndpointAnnotation.cs

```diff
@@ -44,9 +44,23 @@ public EndpointAnnotation(ProtocolType protocol, string? uriScheme = null, strin
         _transport = transport;
         Name = name;
         Port = port;
-        TargetPort = targetPort ?? port;
+
+        TargetPort = targetPort;
+
+        // If the target port was not explicitly set and the service is not being proxied,
+        // we can set the target port to the port.
+        if (TargetPort is null && !isProxied)
+        {
+            TargetPort = port;
+        }
+
         IsExternal = isExternal ?? false;
         IsProxied = isProxied;
+
+        if (TargetPort is int && Port is int && TargetPort == Port && IsProxied)
+        {
+            throw new ArgumentException($"When {nameof(isProxied)} is true, {nameof(targetPort)} must not be equal to {nameof(port)}.");
```

Comment:
This unfortunately depends on what this annotation is applied to. It will work for a singleton container; in fact, that is how we ran in Aspire P1 timeframe when there were no proxyless services. So this check will probably have to be moved to ApplicationExecutor.